src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs

```diff
@@ -1140,9 +1142,278 @@ private void CheckRestrictedTypeReceiver(BoundExpression expression, CSharpCompi
 
             return new BoundCall(node, receiver, method, args, argNames, argRefKinds, isDelegateCall: isDelegateCall,
                         expanded: expanded, invokedAsExtensionMethod: invokedAsExtensionMethod,
-                        argsToParamsOpt: argsToParams, resultKind: LookupResultKind.Viable, binderOpt: this, type: returnType, hasErrors: gotError);
+                        argsToParamsOpt: argsToParams, defaultArguments, resultKind: LookupResultKind.Viable, binderOpt: this, type: returnType, hasErrors: gotError);
+        }
+
+#nullable enable
+
+        private static SourceLocation GetCallerLocation(SyntaxNode syntax)
+        {
+            var token = syntax switch
+            {
+                InvocationExpressionSyntax invocation => invocation.ArgumentList.OpenParenToken,
+                ObjectCreationExpressionSyntax objectCreation => objectCreation.NewKeyword,
+                ConstructorInitializerSyntax constructorInitializer => constructorInitializer.ArgumentList.OpenParenToken,
+                PrimaryConstructorBaseTypeSyntax primaryConstructorBaseType => primaryConstructorBaseType.ArgumentList.OpenParenToken,
+                ElementAccessExpressionSyntax elementAccess => elementAccess.ArgumentList.OpenBracketToken,
+                _ => syntax.GetFirstToken()
+            };
+
+            return new SourceLocation(token);
         }
 
+        internal BoundExpression BindDefaultArgument(SyntaxNode syntax, ParameterSymbol parameter, Symbol containingMember, bool enableCallerInfo, DiagnosticBag diagnostics)
+        {
+            Debug.Assert(parameter.IsOptional);
+
+            TypeSymbol parameterType = parameter.Type;
+            if (Flags.Includes(BinderFlags.ParameterDefaultValue))
+            {
+                // This is only expected to occur in recursive error scenarios, for example: `object F(object param = F()) { }`
+                // We return a non-error expression here to ensure ERR_DefaultValueMustBeConstant (or another appropriate diagnostics) is produced by the caller.
+                return new BoundDefaultExpression(syntax, parameterType) { WasCompilerGenerated = true };
```

Comment:
I this could technically change observable behavior for nullability warnings or IOperation results, but I think that's fine.

Reply:
If we decide to start giving nullability warnings for default arguments then we might want to add a nullable suppression to this node, but we'll cross that bridge if we come to it